[Workflow] Add purge/pause/resume/raise event to docs (#3332)
dapr/docs: daprdocs/content/en/developing-applications/building-blocks/workflow/howto-manage-workflow.md

Dictionary<string, string> workflowOptions; // This is an optional parameter

> CancellationToken cts = CancellationToken.None;

Reviewer: Let's remove this `cts` variable and all references to it below. I've made a similar change already in [this PR](https://github.com/dapr/dotnet-sdk/pull/1087/files#diff-e3110f048c2796ea0bc661b963ead4b5f91d27ca1cffea9929ea614b8b8857cf).

> // Start the workflow. This returns back a "StartWorkflowResponse" which contains the instanceID for the particular workflow instance.

Reviewer: ```suggestion
// Start the workflow. This returns back a "StartWorkflowResponse" which contains the instance ID for the particular workflow instance.
```

StartWorkflowResponse startResponse = await daprClient.StartWorkflowAsync(orderId, workflowComponent, workflowName, input, workflowOptions, cts);

> // Get information on the workflow. This response will contain information such as the status of the workflow, when it started, and more!

Reviewer 2: ```suggestion
// Get information on the workflow. This response contains information such as the status of the workflow, when it started, and more!
```

GetWorkflowResponse getResponse = await daprClient.GetWorkflowAsync(orderId, workflowComponent, workflowName);

// Terminate the workflow
await daprClient.TerminateWorkflowAsync(orderId, workflowComponent, cts);

// Raise an event (an incoming purchase order) that your workflow will wait for. This returns the item waiting to be purchased.
await daprClient.RaiseWorkflowEventAsync(orderId, workflowComponent, workflowName, input, cts);

// Pause

> await daprClient.PauseWorkflowAsync(orderId, workflowComponent, cts);

Reviewer 2: I saw a discussion and code that said this was going to be called suspend, instead of pause. This code https://github.com/dapr/dapr/pull/6106/files. Is that correct?

Author: @RyanLettieri @cgillum can you confirm?

Reviewer 3: It was my understanding that the code was going to reflect Pause/Resume rather than Suspend/Resume. However, the value returned to the user for the status when they call GET on a workflow instance that has been paused is 'Suspended'. I think we should leave it as is for now and potentially revisit later on.

Reviewer: Unfortunately, there's an inconsistency which I think we missed in the implementation. The APIs are all called "Pause" but the workflow status is changed to "Suspended", as Ryan mentioned.

Reviewer 2: Ok. So can we fix this issue in 1.11?

Reviewer: We could change the `GetWorkflow` API handler to return "PAUSED" instead of "SUSPENDED". I think we'd need to take this up with the Dapr maintainers as I expect there will be some angst about making changes this late in the milestone.

GetWorkflowResponse getResponse = await daprClient.GetWorkflowAsync(orderId, workflowComponent, workflowName);

// Terminate the workflow

> GetWorkflowResponse getResponse = await daprClient.TerminateWorkflowAsync(orderId, workflowComponent, cts);

Reviewer 3: There is no response object returned from the terminate call. It should look like this:
await daprClient.TerminateWorkflowAsync(orderId, workflowComponent, cts);

This applies to the RaiseEvent, Pause, Resume, and Purge calls below as well

POST http://localhost:3500/v1.0-alpha1/workflows/dapr/12345678/terminate
```

> ### Pause or resume a workflow

Reviewer 3: @cgillum Correct me if I'm wrong, but for pause and resume we no longer need workflow names, right? So the pause and resume calls below can omit the "OrderProcessingWorkflow" and simply use the workflow component and instance ID. In fact, it looks like you already have that in the document below.

To terminate your workflow with an ID `12345678`, run:

```http
POST http://localhost:3500/v1.0-alpha1/workflows/dapr/12345678/terminate
```

### Raise an event

For workflow components that support subscribing to external events, such as the Dapr Workflow engine, you can use the following "raise event" API to deliver a named event to a specific workflow instance.

```http
POST http://localhost:3500/v1.0-alpha1/workflows/<workflowComponentName>/<instanceID>/raiseEvent/<eventName>
```

> An `eventName` can be any function. 

### Pause or resume a workflow

> To plan for down-time, wait for inputs, and more, you can pause and then resume a workflow. To stop a workflow with an ID `12345678` until triggered to resume, run:

Reviewer 2: ```suggestion
To plan for down-time, wait for inputs, and more, you can pause and then resume a workflow. To pause a workflow with an ID `12345678` until triggered to resume, run:
```

```http
POST http://localhost:3500/v1.0-alpha1/workflows/dapr/12345678/pause
```

To resume a workflow with an ID `12345678`, run:

```http
POST http://localhost:3500/v1.0-alpha1/workflows/dapr/12345678/resume
```

### Purge a workflow 

The purge API can be used to permanently delete workflow metadata from the underlying state store, including any stored inputs, outputs, and workflow history records. This is often useful for implementing data retention policies and for freeing resources.

> Only workflow instances in the COMPLETED, FAILED, or TERMINATED state can be purged. If the workflow is in any other state, calling purge will return an error.

Reviewer 2: ```suggestion
Only workflow instances in the COMPLETED, FAILED, or TERMINATED state can be purged. If the workflow is in any other state, calling purge returns an error.
```

POST http://localhost:3500/v1.0-alpha1/workflows/dapr/12345678/terminate
```

### Raise an event

> Run the following for your workflow to create "wait for external event" tasks that:

Reviewer 2: I still do not understand this. Is there an API to register for an event and an API to raise and event?
Can you provide a complete example of registering for an event X and then the workflow being triggered by event X happening? Reading this, it seems raise event is more like registered for an event. Is that correct, or am  I misunderstanding this?

Author: @RyanLettieri any suggestions for what is the best way to demonstrate the HTTP raise event call better?

Reviewer 3: Hmm I don't think we can actually register an event through the HTTP api, only call an already registered event. Just to double check, @cgillum , is this correct?

Reviewer: This reminds me that we need to update the workflow patterns document to describe the human interaction pattern, which is where external events comes into play. Until then, the best docs we have right now are [here](https://docs.dapr.io/developing-applications/building-blocks/workflow/workflow-features-concepts/#external-events).

To answer the question, registering interest in an event is an engine-specific action. For example, in Dapr Workflow, your workflow can call a `WaitForExternalEvent` API to register interest in an event. Temporal has something similar. We don't have a common interface for registering interest in events - only a public API for raising the events.

Reviewer 2: Still confused here. between sending and registering for events 
 
So there is
1)A  language SDK specific way to register for events with `WaitForExternalEvent`you mention above ^^

2) And also there is this API
// Raise an event (an incoming purchase order) that your workflow will wait for. This returns the item waiting to be purchased.
await daprClient.RaiseWorkflowEventAsync(orderId, workflowComponent, workflowName, input, cts);

3) And then there is an HTTP way to send an event, but then says below that this will "Subscribe to external events" and "Await those tasks to block execution until the event is received.", so that description is wrong then?

Reviewer: Oh, sorry, I wasn't paying close enough attention to the context of your question. Yes, this description is not accurate - there's nothing about this API that will do any kind of subscribing or awaiting. Only the workflow itself can do that.

Here's my proposal for how to describe this API:

*For workflow components that support subscribing to external events, such as the Dapr Workflow engine, you can use the following "raise event" API to deliver a named event to a specific workflow instance.*

```http
POST http://localhost:3500/v1.0-alpha1/workflows/<workflowComponentName>/<instanceID>/raiseEvent/<eventName>
```

Reviewer 2: Yes, this is better.

1. Subscribe to external events.

> 1. Await those tasks to block execution until the event is received.

Reviewer 3: Nit: Should be a 2 instead of a 1

Author: In markdown, using `1.` for all numbers in a numbered list translates to `1.`, `2.`, `3.`, etc. once published, without you having to remember to renumber the list if you add steps later.

```bash
POST http://localhost:3500/v1.0-alpha1/workflows/<workflowComponentName>/<instanceID>/raiseEvent/<eventName>
```

> An `eventName` can be any function. 

> ### Pause or resume a workflow

Reviewer 2: Suspend instead of Pause? Which term will be used?

```suggestion
### Pause or resume a workflow
```

Reviewer: Pause is the term we're using.

To plan for down-time, wait for inputs, and more, you can pause and then resume a workflow. To stop a workflow with an ID `12345678` until triggered to resume, run:

```bash
POST http://localhost:3500/v1.0-alpha1/workflows/dapr/12345678/pause
```

To resume a workflow with an ID `12345678`, run:

```bash
POST http://localhost:3500/v1.0-alpha1/workflows/dapr/12345678/resume
```

### Purge a workflow 

> To remove actor state from your state store and avoid overloading your state stores, run:

Reviewer 2: Actors are an implementation detail and I suggest just saying remove the state. 
```suggestion
To remove all the workflow and activity state from your state store to reduce storage use, run:
```
Also when should you call purge? What is the recommendation, once a workflow has completed? SInce there is a danger here that if you call this whilst a workflow is running, won't this invalidate the workflow? What is the guidance here? Is there a "Danger: Be careful when calling purge API" warning?

Author: @RyanLettieri could you elaborate the risks of calling purge?

Reviewer 3: If the current runtime state is not completed, it is not able to be purged and will return an error stating "orchestration has not yet completed". Purge should be called on completed workflow instances to free up potentially non-needed items in the statestore.

Reviewer 2: I added a comment to add this detail

```bash

> POST http://localhost:3500/v1.0-alpha1/workflows/dapr/OrderProcessingWorkflow/purge

Reviewer 3: "OrderProcessingWorkflow" should be "12345678" since we purge based on the instance ID, not the workflow name.

POST http://localhost:3500/v1.0-alpha1/workflows/dapr/12345678/terminate
```

### Raise an event

Run the following for your workflow to create "wait for external event" tasks that:

> 1. Subscribe to external events.

Reviewer 2: ```suggestion
1. Workflow is registered for an external event  to happen in code
```

1. Await those tasks to block execution until the event is received.

```bash
POST http://localhost:3500/v1.0-alpha1/workflows/<workflowComponentName>/<instanceID>/raiseEvent/<eventName>
```

> An `eventName` can be any function. 

### Pause or resume a workflow

To plan for down-time, wait for inputs, and more, you can pause and then resume a workflow. To stop a workflow with an ID `12345678` until triggered to resume, run:

> ```bash

Reviewer: Instead of `bash` I think it would be more appropriate to use `http` for all these.

POST http://localhost:3500/v1.0-alpha1/workflows/dapr/12345678/pause
```

To resume a workflow with an ID `12345678`, run:

```bash
POST http://localhost:3500/v1.0-alpha1/workflows/dapr/12345678/resume
```

### Purge a workflow 

To remove all the workflow and activity state from your state store to reduce storage use, run:

```bash
POST http://localhost:3500/v1.0-alpha1/workflows/dapr/12345678/purge
```

> 

Reviewer 2: ```suggestion
If the workflow state is not anything other than "Completed" calling `purge` returns an error stating "Workflow has not yet completed". Purge should only be called on completed workflow instances to free resources.
```

Reviewer: I described this a bit more explicitly in the .NET SDK. How about this?

> The purge API can be used to permanently delete workflow metadata from the underlying state store,
> including any stored inputs, outputs, and workflow history records. This is often useful for implementing
> data retention policies and for freeing resources.

> Only workflow instances in the COMPLETED, FAILED, or TERMINATED state can be purged. If the workflow is in any other state, calling `purge` will return an error.

Reviewer 2: Yes, this is much better, please update this.
